src/sentry/api/authentication.py

```diff
@@ -138,6 +144,26 @@ def authenticate_credentials(self, token):
             raise AuthenticationFailed('UserApplication inactive or deleted')
 
         with configure_scope() as scope:
+            scope.set_tag("api_token_type", self.token_name)
             scope.set_tag("api_token", token.id)
 
         return (token.user, token)
+
+
+class DSNAuthentication(StandardAuthentication):
+    token_name = b'dsn'
+
+    def authenticate_credentials(self, token):
+        try:
+            key = ProjectKey.from_dsn(token)
```

Comment:
@mitsuhiko from_dsn ignores the secret_key if present

Reply:
Sorry I get what you mean but the DSN is somewhat easier since then you can just fully transpose it. It also includes the project ID which makes brute forcing a bit more complex.

Reply:
I don't disagree but we use `sentry_key` (PK only) for CSP and minidumps already (project ID is in the payload) and the loader cdn url is just the public key. (eg: https://js.sentry-cdn.com/3bf79c99b89f4725b41540831f92b03d.min.js).

So at least we started going down the path of PK only being a thing.

Reply:
Yeah but those are still DSN based URLs. We could always add public key support but in the use cases I’m imagining they aren’t using minidumps or csp endpoints.

Reply:
What if we additionally add something like:

```python
...
except ProjectKey.DoesNotExist:
    try:
        key = ProjectKey.objects.get_from_cache(
            public_key=token
         )
    except ProjectKey.DoesNotExist:
        raise AuthenticationFailed('Invalid token')
```

I don't really have a strong opinion about this but this would simply enable both.

Reply:
Are we accepting public keys without project IDs now? I thought we always required public key + project ID (which is also why the DSN is nice).

Reply:
@dcramer we do in some situations or rather the code is structured in a way where we already require the public key to be unique. So for instance relay's auth will only use the public key to manage the state associated with the key, and then check it against the project at a later point. Likewise the loader URLs only include the public key, not the project key.

Reply:
(The reason why relay auth effectively has to work this way is because auth happens independent of the actual URL you hit upfront. So only at a later point where we have the auth info we can check against the resource you are hitting which might have the project ID in it)

Reply:
So are you all fine just accepting the DSN for this then? I dont think it makes sense to add more forms of auth unless we're using it, and we can always improve it in the future.

Reply:
I just don't understand why we're not just making it `Authorization: PublicKey <public-key>` instead of the full DSN. Just having the public key is going to be easier to handle going forward.

Reply:
To be clear: having a full DSN in the auth header makes absolutely no sense if we ever need to support it in the relay as we would need to discard everything but the public key.

Reply:
@mitsuhiko imagine SENTRY_DSN- you only set the full DSN. Yeah our SDK parses it out, but the user configures the DSN. Now you want to send feedback (but not via the JS SDK). You already have the DSN so being able to just submit something using the exact same config value is a lot less effort. Otherwise you’re effectively setting another key.